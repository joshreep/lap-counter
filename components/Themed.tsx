/**
 * Learn more about Light and Dark modes:
 * https://docs.expo.io/guides/color-schemes/
 */

import { Text as DefaultText, View as DefaultView, TextInput as DefaultTextInput } from 'react-native'

import Colors from '@/constants/Colors'
import { useColorScheme } from './useColorScheme'
import styled from 'styled-components/native'

type ThemeProps = {
  lightColor?: string
  darkColor?: string
}

export type TextProps = ThemeProps & DefaultText['props']
export type ViewProps = ThemeProps & DefaultView['props']
export type TextInputProps = ThemeProps & DefaultTextInput['props']

export function useThemeColor(
  props: { light?: string; dark?: string },
  colorName: keyof typeof Colors.light & keyof typeof Colors.dark,
) {
  const theme = useColorScheme() ?? 'light'
  const colorFromProps = props[theme]

  if (colorFromProps) {
    return colorFromProps
  } else {
    return Colors[theme][colorName]
  }
}

export const TextInput = styled(DefaultTextInput)`
  color: ${({ theme }) => theme.colors.text};
  background-color: ${({ theme }) => theme.colors.inputBackground};
  border-radius: 4px;
`

export const Text = styled(DefaultText)`
  color: ${({ theme }) => theme.colors.text};
`

export const View = styled(({ lightColor, darkColor, ...rest }: ViewProps) => <DefaultView {...rest} />)`
  background-color: ${(props) =>
    (props.theme.dark ? props.darkColor : props.lightColor) ?? props.theme.colors.background};
`
