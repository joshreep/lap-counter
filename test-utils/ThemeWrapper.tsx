import SCDarkTheme from '@/theme/dark-theme'
import SCLightTheme from '@/theme/light-theme'
import {
  DarkTheme as RNDarkTheme,
  DefaultTheme as RNDefaultTheme,
  ThemeProvider as RNThemeProvider,
} from '@react-navigation/native'
import { FC, PropsWithChildren } from 'react'
import { ColorSchemeName } from 'react-native'
import { ThemeProvider as SCThemeProvider } from 'styled-components/native'

export function getThemeWrapper(theme: ColorSchemeName) {
  const Wrapper: FC<PropsWithChildren> = ({ children }) => {
    return (
      <RNThemeProvider value={theme === 'dark' ? RNDarkTheme : RNDefaultTheme}>
        <SCThemeProvider theme={theme === 'dark' ? SCDarkTheme : SCLightTheme}>{children}</SCThemeProvider>
      </RNThemeProvider>
    )
  }

  return Wrapper
}
