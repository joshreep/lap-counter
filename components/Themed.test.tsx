import { getThemeWrapper } from '@/test-utils/ThemeWrapper'
import { render, screen } from '@testing-library/react-native'
import { ColorSchemeName } from 'react-native'
import { Text, TextInput, View } from './Themed'

const themes: ColorSchemeName[] = ['dark', 'light']

themes.forEach((theme) => {
  const wrapper = getThemeWrapper(theme)

  describe(`TextInput: ${theme} theme`, () => {
    test('should match snapshot', () => {
      render(<TextInput value="Check, check. Is this thing on?" />, { wrapper })
      expect(screen.toJSON()).toMatchSnapshot()
    })
  })

  describe(`Text: ${theme} theme`, () => {
    test('should match snapshot', () => {
      render(<Text>Hello there</Text>, { wrapper })
      expect(screen.toJSON()).toMatchSnapshot()
    })
  })

  describe(`View: ${theme} theme`, () => {
    test('should match snapshot', () => {
      render(
        <View>
          <Text>Hello there</Text>
        </View>,
        { wrapper },
      )
      expect(screen.toJSON()).toMatchSnapshot()
    })
  })
})

describe('TextInput', () => {})
