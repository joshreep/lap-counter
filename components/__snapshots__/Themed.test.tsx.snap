// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text: dark theme should match snapshot 1`] = `
<Text
  style={
    {
      "color": "rgb(255, 255, 255)",
    }
  }
>
  Hello there
</Text>
`;

exports[`Text: light theme should match snapshot 1`] = `
<Text
  style={
    {
      "color": "#111",
    }
  }
>
  Hello there
</Text>
`;

exports[`TextInput: dark theme should match snapshot 1`] = `
<TextInput
  style={
    {
      "backgroundColor": "rgba(255,255,255, .1)",
      "borderBottomLeftRadius": 4,
      "borderBottomRightRadius": 4,
      "borderTopLeftRadius": 4,
      "borderTopRightRadius": 4,
      "color": "rgb(255, 255, 255)",
    }
  }
  value="Check, check. Is this thing on?"
/>
`;

exports[`TextInput: light theme should match snapshot 1`] = `
<TextInput
  style={
    {
      "backgroundColor": "rgba(0, 0, 0, .07)",
      "borderBottomLeftRadius": 4,
      "borderBottomRightRadius": 4,
      "borderTopLeftRadius": 4,
      "borderTopRightRadius": 4,
      "color": "#111",
    }
  }
  value="Check, check. Is this thing on?"
/>
`;

exports[`View: dark theme should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(255, 255, 255, .07)",
    }
  }
>
  <Text
    style={
      {
        "color": "rgb(255, 255, 255)",
      }
    }
  >
    Hello there
  </Text>
</View>
`;

exports[`View: light theme should match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "rgba(0, 0, 0, .05)",
    }
  }
>
  <Text
    style={
      {
        "color": "#111",
      }
    }
  >
    Hello there
  </Text>
</View>
`;
